# Ước lượng Pseudo Ground Truth Transfer Entropy

Notebook này được dùng để chạy KSG trên N rất lớn (200,000) nhằm tính giá trị transfer entropy tham chiếu cho các bộ dữ liệu không có công thức đóng (VAR nonlinear và Periodic coupling).

In [ ]:
import numpy as np
import yaml
import warnings
warnings.filterwarnings("ignore")

from pqrst.data.synthetic.var_nonlinear import generate_var_nonlinear
from pqrst.data.synthetic.periodic_coupling import generate_periodic_coupling
from pqrst.baselines.ksg import KSGTEEstimator

np.random.seed(42)

## Khởi tạo Estimator

In [ ]:
est = KSGTEEstimator(k=4)
N_SAMPLES = 200000

## VAR Nonlinear

In [1]:
x_c, y_c = generate_var_nonlinear(N_SAMPLES, a=0.5, b=0.5, c=0.6, noise_std=0.5, seed=42)
te_c = est.estimate(x_c, y_c)
print(f"VAR Nonlinear (Coupled): {te_c:.6f} nats")

x_u, y_u = generate_var_nonlinear(N_SAMPLES, a=0.5, b=0.5, c=0.0, noise_std=0.5, seed=42)
te_u = est.estimate(x_u, y_u)
print(f"VAR Nonlinear (Uncoupled): {te_u:.6f} nats")

VAR Nonlinear (Coupled): 0.124832 nats
VAR Nonlinear (Uncoupled): -0.000880 nats


## Ghi chu ve gia tri "uncoupled"

KSG o N=200,000 cho ra TE(uncoupled) khong dung bang 0 tuyet doi (bias huu han mau: VAR Nonlinear = -0.000880, Periodic Coupling = -0.002498) - day la bias am dien hinh cua KSG khi TE that gan 0. Vi cac cau hinh "uncoupled" duoc dung `c=0` / `coupling_k=0` (khong ghep noi theo dung thiet ke mo hinh), TE that ve mat ly thuyet BANG 0 CHINH XAC (2 qua trinh doc lap). Vi vay `pseudo_ground_truth_te` trong configs/synthetic/*.yaml duoc ghi la `0.0` (gia tri ly thuyet dung), KHONG phai gia tri KSG in ra o tren - gia tri KSG chi dung de doi chieu/audit.

## Periodic Coupling

In [2]:
x_pc, y_pc = generate_periodic_coupling(N_SAMPLES, omega_x=0.3, omega_y=0.31, coupling_k=0.8, noise_std=0.1, seed=42)
te_pc = est.estimate(x_pc, y_pc)
print(f"Periodic Coupling (Coupled): {te_pc:.6f} nats")

x_pu, y_pu = generate_periodic_coupling(N_SAMPLES, omega_x=0.3, omega_y=0.31, coupling_k=0.0, noise_std=0.1, seed=42)
te_pu = est.estimate(x_pu, y_pu)
print(f"Periodic Coupling (Uncoupled): {te_pu:.6f} nats")

Periodic Coupling (Coupled): 0.324613 nats
Periodic Coupling (Uncoupled): -0.002498 nats
